Add support for bulleted lists and underline

eschramm opened the pull request:
A simplified implementation of bulleted (non-ordered) lists.
Current supports 3 styles of bullets:
- filled circle `•` for first-level indent
- open circle `◦` for second-level indent
- filled square `∙` for third-level and beyond indent

Indent follows Markdown specifications of 4-spaces per level of indent. This also uses a font override for the bullets as the standard system font had smaller bullets than normal and I thought they looked funny. The indent uses raw `NSParagraphStyle` formatting of indentation.

As in the Android implementation, they look a little funny when center-justified. Ideally style should use `Left` justification to make it look better

Supported Markdown bullets are `-`, `*`, and `+`. The parser looks for a `{bullet}{space}` to mark the start of a list item and then verifies this is immediately preceded by whitespace and then a new line or is the beginning of the string. Otherwise text like "this - may" parse as an inline bullet.

While I was in there, I took a stab at adding plain underline which is not included in standard Markdown spec but has been requested by at least 2 studies. This implementation simply uses a triple-underscore `___This is underlined___` wrapper. The wrapper elements are a constant, so this can be easily changed if we decide on a different string set.

Here's an example to demonstrate with JSON below.

![Simulator Screen Shot - iPhone 11 - 2020-10-08 at 20 29 32](https://user-images.githubusercontent.com/1008462/95530993-fdaa1200-09a4-11eb-83ab-6b9177082202.png)

```json
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "styles": {
    "type": "Style",
    "textAlignment": "Left",
    "detailTextAlignment": "Left"
  },
  "steps": [
    {
      "type": "InstructionStep",
      "text": "This is plain - text - and * these + should ___NOT___ trigger bulleting. ___**Bold+Underlined**___ ipsum dolor sit amet. \\___Escaped Underline\\___.\n- bullet\n    - indented\n    - indented same but this time with some significant text to ensure word wrapping occurs. Check for appropriate indentation of subsequent lines. This also should handle auto-rotation.\n- outdent\n- another\n    - one level\n        - two levels\n            - three levels\n        - two levels (2) - more text for word wrapping example. Four spaces per indent level.\n            - three levels again\n                - ___four___ levels\n\\- escaped bullet",
      "detailText": "- this should also parse as a bullet since this is the beginning of ___detailText___ and that qualifies as a valid bullet in addition to following a new line",
      "title": "This is the ___Title___",
      "identifier": "New Step 1",
    }
  ]
}
```

eschramm committed:
add bulleted list support

eschramm committed:
add underline support

eschramm committed:
add missing break to switch statement

eschramm committed:
ensure bullets start with new line or start of string

mmertsock commented on NSAttributedString+Markdown.m:
```diff
@@ -541,6 +571,74 @@ static void updateAttributedString(NSMutableAttributedString *result, NSString *
+                                    break;
+                            }
+                    
+                            UIFont *baseFont = [result attribute:NSFontAttributeName atIndex:(beginRange.location - mutationOffset) effectiveRange:NULL];
```
Replace UIFont with `FONT_CLASS`

eschramm replied:
Good catch

mmertsock commented on NSAttributedString+Markdown.m:
```diff
@@ -541,6 +571,74 @@ static void updateAttributedString(NSMutableAttributedString *result, NSString *
+                            [result addAttribute:NSFontAttributeName value:updatedBulletFont range:NSMakeRange(mutatedBeginRange.location, 1)];
+                            
+                            NSRange listItemRangeWithBullet = NSMakeRange(mutatedBeginRange.location, endRange.location - mutatedBeginRange.location + endRange.length - mutationOffset - 1);
+                            NSMutableParagraphStyle *paragraphStyle = [[result attribute:NSParagraphStyleAttributeName atIndex:listItemRangeWithBullet.location effectiveRange:NULL] mutableCopy];
```
Playing around with the markdown sample app built for macOS, sometimes paragraphStyle is nil, which causes an objective-c exception in the `[result addAttribute:value:range:]` call below. Perhaps it's a Mac-only issue.

I was able to hack around the exception by checking for nil paragraphStyle here and using `[[NSParagraphStyle defaultParagraphStyle] mutableCopy]` — not sure if that's the best solution though.

Anyway, in our ResearchKit code should we catch any objective-c exceptions when calling `initWithMarkdownRepresentation` to prevent crashes?

mmertsock commented on NSAttributedString+Markdown.m:
```diff
@@ -541,6 +571,74 @@ static void updateAttributedString(NSMutableAttributedString *result, NSString *
+                                    break;
+                                }
+                            }
+                            NSUInteger bulletLevel = spacesBeforeBullet / 4;
```
Would be nice to define the `4` value as a constant at the top of the file

eschramm replied:
Agreed

mmertsock commented on NSAttributedString+Markdown.m:
```diff
@@ -541,6 +571,74 @@ static void updateAttributedString(NSMutableAttributedString *result, NSString *
+#endif
+                            break;
+                        }
+                        case MarkdownSpanUnderline: {
```
Would be easier to read/review if the `MarkdownSpanUnderline` case were after the MarkdownSpanEmphasisDouble case, since the implementation is identical

eschramm replied:
Good idea!

mmertsock commented on NSAttributedString+Markdown.m:
```diff
@@ -541,6 +571,74 @@ static void updateAttributedString(NSMutableAttributedString *result, NSString *
+                            [result addAttribute:NSFontAttributeName value:updatedBulletFont range:NSMakeRange(mutatedBeginRange.location, 1)];
+                            
+                            NSRange listItemRangeWithBullet = NSMakeRange(mutatedBeginRange.location, endRange.location - mutatedBeginRange.location + endRange.length - mutationOffset - 1);
+                            NSMutableParagraphStyle *paragraphStyle = [[result attribute:NSParagraphStyleAttributeName atIndex:listItemRangeWithBullet.location effectiveRange:NULL] mutableCopy];
```
Playing around with the markdown sample app built for macOS, sometimes paragraphStyle is nil, which causes an objective-c exception in the `[result addAttribute:value:range:]` call below. Perhaps it's a Mac-only issue.

I was able to hack around the exception by checking for nil paragraphStyle here and using `[[NSParagraphStyle defaultParagraphStyle] mutableCopy]` — not sure if that's the best solution though.

Anyway, in our ResearchKit code should we catch any objective-c exceptions when calling `initWithMarkdownRepresentation` to prevent crashes?

eschramm replied:
I'm glad you caught that. However, catching the exception is tricky. At first I thought, let's use the `ObjCtoSwiftExceptionHandler` in CEVMobile and wrap the presentation of the `UITaskViewController`. But since this exception is thrown during the `viewDidAppear` chain of events, it cannot be caught that way. Instead, I think your idea of ensuring paragraphStyle isn't nil seems acceptable to me because if for some reason, there is no formatting, I think it would be fine to use whatever the default formatting would be to append our indent/tab formatting to.

mmertsock replied:
sounds good. the nil paragraphStyle thing I encountered may be a quirk of macOS or the sample app anyway

mmertsock reviewed (approved):
Nice! Just minor comments

eschramm committed:
PR fixes

- change spacesPerBulletIndentLevel to a constant
- use FONT_CLASS vs direct UIFont
- move MarkdownSpanUnderline to MarkdownSpanEmphasisDouble to save duplication
- guard against NSAttributedString having no paragraphStyle

mmertsock reviewed (approved):
LGTM 🚢
